### PR TITLE
enhancement(kafka sink): Add acknowledgements support

### LINKED ISSUE #7656: Add support for acknowledgements to `kafka` sink
Ref #7337

### PR DESCRIPTION
**Pull request by bruceg**
Closes #7656 

I could not figure a good way of forcing the integration tests to produce an error other than providing incorrect authentication credentials, which just caused a very long timeout loop, not good for automated testing. As such the error path is untested.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

### CODE REVIEW

File: src/sinks/kafka.rs [hunk 1]
```diff
@@ -338,13 +348,17 @@ impl Sink<Event> for KafkaSink {
...
-                Some((seqno, Ok(result))) => {
+                Some((seqno, Ok(result), metadata)) => {
                     match result {
                         Ok((partition, offset)) => {
-                            trace!(message = "Produced message.", ?partition, ?offset)
+                            metadata.update_status(EventStatus::Delivered);
+                            trace!(message = "Produced message.", ?partition, ?offset);
+                        }
+                        Err(error) => {
+                            metadata.update_status(EventStatus::Errored); // should this be `Failed`?
```
Comment (lukesteensen) [inline src/sinks/kafka.rs:358]: If `Errored` implies a problem with the data and `Failed` implies a problem with the server, this should probably be `Failed`. Kafka messages are just bytes so it's hard to make one invalid.

File: src/sinks/kafka.rs [hunk 2]
```diff
@@ -405,7 +420,7 @@ fn encode_event(
     mut event: Event,
     key_field: &Option<String>,
     encoding: &EncodingConfig<Encoding>,
-) -> (Vec<u8>, Vec<u8>) {
```
Comment (lukesteensen) [inline src/sinks/kafka.rs:408]: It'd be nice if `encode_event` didn't need to worry about the metadata. Would it make sense to have something like the `into_parts` method to split out the metadata before encoding? Or even something like `take_metadata`?

I realize it's probably somewhat complicated here by working on both logs and metrics, but ideally that wouldn't matter.

**Review by lukesteensen — APPROVED** [review]
One small note about what feels like some extra plumbing of metadata, but overall this looks reasonable to me.

File: src/sinks/kafka.rs (hunk 1 above)
Comment (bruceg) [reply src/sinks/kafka.rs:358]: It's more an issue of temporary vs permanent errors. That is, `Errored` indicates the events may be retried but `Failed` indicates permanent rejection.

File: src/sinks/kafka.rs (hunk 2 above)
Comment (bruceg) [reply src/sinks/kafka.rs:408]: So, long story here. When I was first developing the initial dummy event metadata, I started with turning `Event` into a new wrapper struct containing the metadata and the actual event enum. This ended up requiring major code changes all over the place, and numerous conversion traits, etc. So, I reworked it to store the event metadata alongside the log and metric data, which required only a relatively few changes.

This unfortunately means that a `fn Event::into_parts` is simply not possible, as `Event` is still an enum with variable parts behind it. `fn Event::take_metadata` is a possibility, calling down to an equivalent function in each variant, which in turn do a `swap` with `EventMetadata::default()`, which doesn't currently allocate any memory or do other real work.

As this goes beyond just the `kafka` sink, I'll open up a separate issue to track this.

File: src/sinks/kafka.rs (hunk 1 above)
Comment (lukesteensen) [reply src/sinks/kafka.rs:358]: In that case I'd probably lean towards `Errored`. I'm not entirely sure what a "permanent rejection" would entail for Kafka (maybe a message that's too large?).

**Pull request merged**